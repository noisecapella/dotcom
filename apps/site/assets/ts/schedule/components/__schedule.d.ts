import { TypedRoutes } from "../../stop/components/__stop";
import {
  Alert,
  Route,
  PredictedOrScheduledTime,
  EnhancedRoute,
  DirectionId,
  Service,
  Schedule,
  RoutePattern,
  Shape,
  Stop,
  ClosedStopInfo
} from "../../__v3api";

export interface EnhancedRoutePattern extends RoutePattern {
  shape_id: string;
  headsign: string;
}

export interface ShapesById {
  [key: string]: Shape;
}

export interface RoutePatternsByDirection {
  [key: string]: EnhancedRoutePattern[];
}

export interface ServiceInSelector extends Service {
  "default_service?": boolean;
}

export interface SchedulePageData {
  connections: TypedRoutes[];
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
  fares: Fare[];
  fare_link: string;
  holidays: Holiday[];
  route: EnhancedRoute;
  schedule_note: ScheduleNote | null;
  services: ServiceInSelector[];
  stops: SimpleStopMap;
  direction_id: DirectionId;
  shape_map: ShapesById;
  route_patterns: RoutePatternsByDirection;
  line_diagram: LineDiagramStop[];
  today: string;
  variant: string | null;
}

interface StopData {
  branch: string | null;
  type: "line" | "merge" | "stop" | "terminus" | null;
}

export interface LineDiagramStop {
  stop_data: StopData[];
  route_stop: RouteStop;
  alerts: Alert[];
}

export interface LineDiagramVehicle {
  id: string;
  headsign: string | null;
  status: "in_transit" | "incoming" | "stopped";
  trip_name: string | null;
}

interface RouteStopRoute extends Route {
  "custom_route?": boolean;
}
export interface RouteStop {
  id: string;
  name: string;
  zone: string | null;
  branch: string | null;
  station_info: Stop & { parent_id: string | null; child_ids: string[] };
  route: RouteStopRoute | null;
  connections: RouteStopRoute[];
  stop_features: string[];
  "is_terminus?": boolean;
  "is_beginning?": boolean;
  closed_stop_info: ClosedStopInfo | null;
}

export interface SimpleStopMap {
  [key: string]: SimpleStop[];
}

export interface ServiceSchedule {
  [key: string]: ServiceScheduleInfo;
}

export interface ScheduleInfo {
  schedules: ScheduleWithFare[];
  duration: string;
  route_pattern_id: string;
}

export interface ScheduleWithFare extends Schedule {
  price: string;
  fare_link: string;
}

export interface ServiceScheduleByTrip {
  [key: string]: ScheduleInfo;
}

export interface ServiceScheduleInfo {
  by_trip: ServiceScheduleByTrip;
  trip_order: string[];
}

export interface ScheduleNote {
  peak_service: string;
  offpeak_service: string;
  exceptions: ServiceException[];
  alternate_text: string | null;
}

export interface ServiceException {
  type: string;
  service: string;
}

export interface SimpleStop {
  id: string;
  name: string;
  is_closed: boolean;
  zone: string | null;
}

export interface SchedulePDF {
  title: string;
  url: string;
}

export interface Fare {
  title: string;
  price: string;
}

export interface Holiday {
  name: string;
  date: string;
}

export interface StopPrediction {
  headsign: string;
  route: Route;
  prediction: PredictedOrScheduledTime;
  train_number: string;
}

export type SelectedOrigin = string | null;
export type SelectedStopId = string | null;
export interface UserInput {
  route: string;
  origin: string;
  date: string;
  direction: DirectionId;
}

export type StoreAction =
  | "INITIALIZE"
  | "CHANGE_DIRECTION"
  | "CHANGE_ORIGIN"
  | "CHANGE_DESTINATION"
  | "OPEN_MODAL"
  | "CLOSE_MODAL";

export type ComponentToRender =
  | "MAIN"
  | "SCHEDULE_NOTE"
  | "SCHEDULE_FINDER"
  | "SCHEDULE_DIRECTION"
  | "";
