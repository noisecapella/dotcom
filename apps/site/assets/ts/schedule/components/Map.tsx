import React, { ReactElement, useRef } from "react";
import Map from "../../leaflet/components/Map";
import getBounds from "../../leaflet/bounds";
import {
  MapData,
  MapMarker as Marker
} from "../../leaflet/components/__mapdata";
import CrowdingPill from "./line-diagram/CrowdingPill";
import useVehicleMarkersChannel from "../../hooks/useVehicleChannel";

interface Props {
  channel: string;
  data: MapData;
  currentShapes: string[];
  currentStops: string[];
}

export const iconOpts = (
  icon: string | null
): {
  icon_size?: [number, number];
  icon_anchor?: [number, number];
} => {
  switch (icon) {
    case null:
      return {};

    case "vehicle-bordered-expanded":
      return {
        icon_size: [18, 18], // eslint-disable-line camelcase
        icon_anchor: [6, 6] // eslint-disable-line camelcase
      };

    case "stop-circle-bordered-expanded":
      return {
        icon_size: [12, 12], // eslint-disable-line camelcase
        icon_anchor: [6, 6] // eslint-disable-line camelcase
      };

    default:
      throw new Error(`unexpected icon type: ${icon}`);
  }
};

const zIndex = (icon: string | null): number | undefined =>
  icon === "vehicle-bordered-expanded" ? 1000 : undefined;

export const updateMarker = (marker: Marker): Marker => ({
  ...marker,
  tooltip: (
    <div>
      {marker.vehicle_crowding && (
        <>
          <CrowdingPill crowding={marker.vehicle_crowding} />
          <br />
        </>
      )}
      {marker.tooltip_text}
    </div>
  ),
  icon_opts: iconOpts(marker.icon), // eslint-disable-line camelcase
  z_index: zIndex(marker.icon) // eslint-disable-line camelcase
});

export const isVehicleMarker = (marker: Marker): boolean =>
  marker.icon ? marker.icon.includes("vehicle") : false;

export interface IdHash {
  [id: string]: true;
}

export const shouldRemoveMarker = (
  id: string | null,
  idHash: IdHash
): boolean => id !== null && idHash[id] === true;

export default ({
  data,
  channel,
  currentShapes,
  currentStops
}: Props): ReactElement<HTMLElement> | null => {
  const state = useVehicleMarkersChannel(channel);
  const stopMarkers = data.stop_markers
    ? data.stop_markers
        .filter(mark => currentStops.includes(mark.id as string))
        .map(marker => updateMarker(marker))
    : [];

  const mapData = {
    ...data,
    polylines: data.polylines.filter(p =>
      currentShapes.some(shape => shape === p.id)
    ),
    markers: state.concat(stopMarkers)
  };
  const bounds = useRef(getBounds(stopMarkers));
  return (
    <div className="m-schedule__map">
      <Map bounds={bounds.current} mapData={mapData} />
    </div>
  );
};
