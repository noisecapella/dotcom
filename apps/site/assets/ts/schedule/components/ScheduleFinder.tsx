import React, { ReactElement } from "react";
import { Route, DirectionId } from "../../__v3api";
import {
  SimpleStopMap,
  RoutePatternsByDirection,
  ServiceInSelector,
  ScheduleNote as ScheduleNoteType,
  SelectedOrigin
} from "./__schedule";
import ScheduleFinderForm from "./schedule-finder/ScheduleFinderForm";
import ScheduleFinderModal, {
  Mode as ModalMode
} from "./schedule-finder/ScheduleFinderModal";
import { getCurrentState, storeHandler } from "../store/ScheduleStore";

interface Props {
  updateURL: (origin: SelectedOrigin, direction?: DirectionId) => void;
  services: ServiceInSelector[];
  directionId: DirectionId;
  route: Route;
  stops: SimpleStopMap;
  routePatternsByDirection: RoutePatternsByDirection;
  today: string;
  scheduleNote: ScheduleNoteType | null;
  changeDirection: (direction: DirectionId) => void;
  selectedOrigin: SelectedOrigin;
  changeOrigin: (origin: SelectedOrigin) => void;
  closeModal: () => void;
  modalMode: ModalMode;
  modalOpen: boolean;
}

const ScheduleFinder = ({
  updateURL,
  directionId,
  route,
  services,
  stops,
  routePatternsByDirection,
  today,
  scheduleNote,
  modalMode,
  selectedOrigin,
  changeDirection,
  changeOrigin,
  modalOpen,
  closeModal
}: Props): ReactElement<HTMLElement> => {
  const openOriginModal = (): void => {
    const currentState = getCurrentState();
    const { modalOpen: modalIsOpen } = currentState;
    if (!modalIsOpen) {
      storeHandler({
        type: "OPEN_MODAL",
        newStoreValues: {
          modalMode: "origin"
        }
      });
    }
  };

  const openScheduleModal = (): void => {
    const currentState = getCurrentState();
    const { modalOpen: modalIsOpen } = currentState;
    if (selectedOrigin !== undefined && !modalIsOpen) {
      storeHandler({
        type: "OPEN_MODAL",
        newStoreValues: {
          modalMode: "schedule"
        }
      });
    }
  };

  const handleOriginSelectClick = (): void => {
    storeHandler({
      type: "OPEN_MODAL",
      newStoreValues: {
        modalMode: "origin"
      }
    });
  };

  return (
    <div className="schedule-finder">
      <ScheduleFinderForm
        onDirectionChange={changeDirection}
        onOriginChange={changeOrigin}
        onOriginSelectClick={openOriginModal}
        onSubmit={openScheduleModal}
        route={route}
        selectedDirection={directionId}
        selectedOrigin={selectedOrigin}
        stopsByDirection={stops}
      />

      {modalOpen && (
        <ScheduleFinderModal
          directionChanged={changeDirection}
          handleOriginSelectClick={handleOriginSelectClick}
          originChanged={changeOrigin}
          route={route}
          routePatternsByDirection={routePatternsByDirection}
          scheduleNote={scheduleNote}
          services={services}
          stops={stops}
          today={today}
          updateURL={updateURL}
        />
      )}
    </div>
  );
};

export default ScheduleFinder;
