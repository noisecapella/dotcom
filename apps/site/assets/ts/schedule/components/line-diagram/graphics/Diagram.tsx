import { max } from "lodash";
import React, { ReactElement } from "react";
import { LineDiagramStop } from "../../__schedule";
import {
  hasBranchLines,
  isBranchTerminusStop,
  areOnDifferentBranchLines,
  diagramWidth,
  isMergeStop
} from "../line-diagram-helpers";
import Line from "./Line";
import Merges from "./Merge";
import Stop from "./Stop";
import { routeToModeName } from "../../../../helpers/css";
import { getCurrentState } from "../../../store/ScheduleStore";
import { DirectionId } from "../../../../__v3api";
import { isAGreenLineRoute } from "../../../../models/route";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";

interface DiagramProps {
  lineDiagram: LineDiagramStop[];
}

const branchingDescription = (lineDiagram: LineDiagramStop[]): string => {
  const mergeStops = lineDiagram.filter(isMergeStop);
  const branchText = mergeStops.map(mergeStop => {
    const mergeStopName = mergeStop.route_stop.name;
    const branchNames = mergeStop.stop_data
      .filter(sd => sd.branch)
      .map(sd => sd.branch);
    return `Use ${mergeStopName} to transfer to trains on ${branchNames.join(
      " or "
    )} branches. `;
  });
  return branchText.join("");
};
const diagramDescription = (
  lineDiagram: LineDiagramStop[],
  direction: DirectionId
): string => {
  const text = `${lineDiagram.length} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = lineDiagram[0].route_stop.route!;
  if (destinations) {
    return `${text} going ${names[direction]} to ${destinations[direction]}`;
  }
  return text;
};

const Diagram = ({
  lineDiagram
}: DiagramProps): ReactElement<HTMLElement> | null => {
  const { selectedDirection } = getCurrentState();
  const width = diagramWidth(
    max(lineDiagram.map(ld => ld.stop_data.length)) || 1
  );

  return (
    <>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        role="img"
        aria-labelledby="diagram-title diagram-desc"
        className={`line-diagram-svg ${routeToModeName(
          lineDiagram[0].route_stop.route!
        )}`}
        width={`${width + 4}px`}
        height="100%"
        style={{ left: BASE_LINE_WIDTH / 2 }}
      >
        <title id="diagram-title">
          Line diagram for{" "}
          {isAGreenLineRoute(lineDiagram[0].route_stop.route!)
            ? "Green Line"
            : lineDiagram[0].route_stop.route!.name}
        </title>
        <desc id="diagram-desc">
          {diagramDescription(lineDiagram, selectedDirection)}
          {hasBranchLines(lineDiagram) && branchingDescription(lineDiagram)}
        </desc>

        <defs>{DiagonalHatchPattern}</defs>
        {/* If there are multiple branches, draw the lines and curves for those */
        hasBranchLines(lineDiagram) && <Merges lineDiagram={lineDiagram} />}

        {/* Draw lines between stops */
        lineDiagram.map((current, idx, self) => {
          if (self[idx + 1]) {
            if (
              isBranchTerminusStop(current) &&
              areOnDifferentBranchLines(current, self[idx + 1])
            )
              return null; // don't draw a line from a branch end to the main line

            return (
              <Line
                key={`${current.route_stop.id}-${self[idx + 1].route_stop.id}`}
                from={current}
                to={self[idx + 1]}
              />
            );
          }
          return null; // this is the end of the line, nothing to draw a line to
        })}

        {/* Draw circles for each stop */
        lineDiagram.map(stop => (
          <Stop key={stop.route_stop.id} stop={stop} />
        ))}
      </svg>
    </>
  );
};

export default Diagram;
