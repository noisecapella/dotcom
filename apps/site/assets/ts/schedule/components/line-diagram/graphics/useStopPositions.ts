import { useDispatch } from "react-redux";
import { useRef, useEffect, useCallback } from "react";
import { LineDiagramStop } from "../../__schedule";
import { isMergeStop } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, BRANCH_SPACING } from "./graphic-helpers";

export interface RefList {
  [key: string]: React.MutableRefObject<HTMLElement | null>;
}

function useStopRef(prev: RefList, stop: LineDiagramStop): RefList {
  return {
    ...prev,
    [stop.route_stop.id]: useRef<HTMLElement>(null)
  };
}

const stopById = (stopId: string): ((stop: LineDiagramStop) => boolean) =>
  // eslint-disable-next-line camelcase
  ({ route_stop }) => stopId === route_stop.id;

const xCoordForStop = (stop: LineDiagramStop): number => {
  if (isMergeStop(stop)) {
    return BASE_LINE_WIDTH + 1;
  }
  // usually just a function of stop_data length
  return BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
};

export default function useStopPositions(
  stops: LineDiagramStop[]
): [RefList, () => void] {
  const stopRefsMap: RefList = stops.reduce(useStopRef, {});
  const dispatchStopCoords = useDispatch();
  const updateAllStops = useCallback((): void => {
    Object.entries(stopRefsMap).forEach(([stopId, ref]) => {
      const x = xCoordForStop(stops.find(stopById(stopId))!);
      let coordinates = null;
      if (ref && ref.current) {
        const { offsetTop, offsetHeight } = ref.current;
        const y = offsetTop + offsetHeight / 2;
        coordinates = [x, y];
      }
      dispatchStopCoords({
        type: "set",
        stop: stopId,
        coords: coordinates
      });
    });
  }, [dispatchStopCoords, stopRefsMap, stops]);

  useEffect(() => {
    updateAllStops();
    window.addEventListener("resize", updateAllStops);
    return () => window.removeEventListener("resize", updateAllStops);
  }, [updateAllStops]);

  return [stopRefsMap, updateAllStops];
}
