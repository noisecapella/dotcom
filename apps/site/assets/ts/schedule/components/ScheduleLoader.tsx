import React, { ReactElement } from "react";
import { connect } from "react-redux";
import { useQueryParams, StringParam } from "use-query-params";
import SchedulePage from "../components/SchedulePage";
import ScheduleNote from "../components/ScheduleNote";
import ScheduleDirection from "../components/ScheduleDirection";
import {
  SchedulePageData,
  SelectedStopId,
  ComponentToRender
} from "../components/__schedule";
import { MapData, StaticMapData } from "../../leaflet/components/__mapdata";
import ScheduleFinder from "../components/ScheduleFinder";
import ScheduleFinderModal from "../components/schedule-finder/ScheduleFinderModal";
import { DirectionId } from "../../__v3api";
import {
  mapStateToProps,
  getCurrentState,
  storeHandler
} from "../store/ScheduleStore";

interface Props {
  schedulePageData: SchedulePageData;
  component: ComponentToRender;
  updateURL: (origin: SelectedStopId, direction?: DirectionId) => void;
}

export const ScheduleLoader = ({
  component,
  schedulePageData,
  updateURL
}: Props): ReactElement<HTMLElement> => {
  const [query] = useQueryParams({
    // eslint-disable-next-line @typescript-eslint/camelcase
    "schedule_finder[direction_id]": StringParam,
    "schedule_finder[origin]": StringParam
  });

  const changeDirection = (direction: DirectionId): void => {
    storeHandler({
      type: "CHANGE_DIRECTION",
      newStoreValues: {
        selectedDirection: direction,
        selectedOrigin: null
      }
    });
  };

  const changeOrigin = (origin: SelectedStopId): void => {
    storeHandler({
      type: "CHANGE_ORIGIN",
      newStoreValues: {
        selectedOrigin: origin
      }
    });
    // reopen modal depending on choice:
    storeHandler({
      type: "OPEN_MODAL",
      newStoreValues: {
        modalMode: origin ? "schedule" : "origin"
      }
    });
  };

  const closeModal = (): void => {
    storeHandler({
      type: "CLOSE_MODAL",
      newStoreValues: {}
    });
    // clear parameters from URL when closing the modal:
    updateURL("");
  };

  React.useEffect(() => {
    // get initial values from the store:
    const currentState = getCurrentState();
    const { selectedDirection, selectedOrigin } = currentState;
    let { modalOpen, modalMode } = currentState;

    let newDirection: DirectionId | undefined;
    let newOrigin: SelectedStopId | undefined;

    // modify the store values in case URL has parameters:
    if (query["schedule_finder[direction_id]"] !== undefined) {
      newDirection = query["schedule_finder[direction_id]"] === "0" ? 0 : 1;
    }
    if (query["schedule_finder[origin]"] !== undefined) {
      newOrigin = query["schedule_finder[origin]"];
    }

    if (newDirection !== undefined && newOrigin !== undefined) {
      modalMode = "schedule";
      modalOpen = true;
    }

    storeHandler({
      type: "INITIALIZE",
      newStoreValues: {
        selectedDirection: newDirection || selectedDirection,
        selectedOrigin: newOrigin || selectedOrigin,
        modalMode,
        modalOpen
      }
    });
    // we disable linting in this next line because we DO want to specify an empty array since we want this piece to run only once
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleOriginSelectClick = (): void => {
    storeHandler({
      type: "OPEN_MODAL",
      newStoreValues: {
        modalMode: "origin"
      }
    });
  };

  const {
    route,
    stops,
    services,
    route_patterns: routePatternsByDirection,
    schedule_note: scheduleNote,
    today,
    line_diagram: lineDiagram,
    shape_map: shapesById,
    variant: initialSelectedRoutePatternId
  } = schedulePageData;

  const currentState = getCurrentState();
  if (!!currentState && Object.keys(currentState).length !== 0) {
    const {
      selectedDirection: currentDirection,
      selectedOrigin,
      modalOpen,
      modalMode
    } = currentState;
    if (component === "MAIN") {
      return (
        <SchedulePage
          updateURL={updateURL}
          schedulePageData={schedulePageData}
          closeModal={closeModal}
          selectedDirection={currentDirection}
          changeDirection={changeDirection}
          selectedOrigin={selectedOrigin}
          changeOrigin={changeOrigin}
          modalOpen={modalOpen}
          modalMode={modalMode}
        />
      );
    }

    if (component === "SCHEDULE_NOTE" && scheduleNote) {
      return (
        <>
          <ScheduleNote
            className="m-schedule-page__schedule-notes--mobile"
            scheduleNote={scheduleNote}
          />
          {modalOpen && (
            <ScheduleFinderModal
              closeModal={closeModal}
              directionChanged={changeDirection}
              initialMode={modalMode}
              initialDirection={currentDirection}
              initialOrigin={selectedOrigin}
              handleOriginSelectClick={handleOriginSelectClick}
              originChanged={changeOrigin}
              route={route}
              routePatternsByDirection={routePatternsByDirection}
              scheduleNote={scheduleNote}
              services={services}
              stops={stops}
              today={today}
              updateURL={updateURL}
            />
          )}
        </>
      );
    }

    if (component === "SCHEDULE_FINDER") {
      return (
        <ScheduleFinder
          updateURL={updateURL}
          route={route}
          stops={stops}
          services={services}
          routePatternsByDirection={routePatternsByDirection}
          today={today}
          scheduleNote={null}
          modalMode={modalMode}
          modalOpen={modalOpen}
          directionId={currentDirection}
          changeDirection={changeDirection}
          selectedOrigin={selectedOrigin}
          changeOrigin={changeOrigin}
          closeModal={closeModal}
        />
      );
    }

    if (component === "SCHEDULE_DIRECTION") {
      let mapData: MapData | undefined;
      const mapDataEl = document.getElementById("js-map-data");
      if (mapDataEl) {
        mapData = JSON.parse(mapDataEl.innerHTML);
      }

      let staticMapData: StaticMapData | undefined;
      const staticDataEl = document.getElementById("static-map-data");
      if (staticDataEl) {
        staticMapData = JSON.parse(staticDataEl.innerHTML);
      }

      return (
        <ScheduleDirection
          directionId={currentDirection}
          route={route}
          routePatternsByDirection={routePatternsByDirection}
          shapesById={shapesById}
          mapData={mapData}
          staticMapData={staticMapData}
          lineDiagram={lineDiagram}
          services={services}
          stops={stops}
          today={today}
          scheduleNote={scheduleNote}
          initialSelectedRoutePatternId={initialSelectedRoutePatternId}
        />
      );
    }
  }
  return <></>;
};

export default connect(mapStateToProps)(ScheduleLoader);
