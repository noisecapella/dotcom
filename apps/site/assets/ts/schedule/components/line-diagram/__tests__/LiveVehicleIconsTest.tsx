import React from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import { cloneDeep } from "lodash";
import { RouteStop, LineDiagramStop } from "../../__schedule";
import LiveVehicles, { VehicleIcons } from "../LiveVehicleIcons";
import { createLineDiagramCoordStore } from "../graphics/graphic-helpers";
import { MapMarker } from "../../../../leaflet/components/__mapdata";
import * as UseVehicleChannel from "../../../../hooks/useVehicleChannel";

// mock the redux state
jest.spyOn(redux, "useSelector").mockImplementation(selector =>
  selector({
    "test-stop": [12, 80]
  })
);
const stop = { id: "test-stop", name: "Test Stop" } as RouteStop;
export const vehicles = [
  {
    id: "vehicle-1",
    vehicle_status: "in_transit",
    tooltip_text: "tooltip text for vehicle 1",
    stop_id: stop.id
  },
  {
    id: "vehicle-2",
    vehicle_status: "incoming",
    tooltip_text: "tooltip text for vehicle 2",
    stop_id: stop.id
  },
  {
    id: "vehicle-3",
    vehicle_status: "stopped",
    tooltip_text: "tooltip text for vehicle 3",
    stop_id: stop.id
  }
] as MapMarker[];

const other_vehicles = cloneDeep(vehicles).map(v => {
  v.stop_id = "test-stop-other";
  return v;
});
const store = createLineDiagramCoordStore([
  { route_stop: stop } as LineDiagramStop
]);

const tooltipText = (wrapper: ReactWrapper) =>
  wrapper.find("[tooltipText]").prop("tooltipText");

describe("VehicleIcons", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <VehicleIcons stop_id={stop.id} vehicles={vehicles} />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders each vehicle", () => {
    const vehicleNodes = wrapper.find(".m-schedule-diagram__vehicle");
    expect(vehicleNodes).toHaveLength(vehicles.length);
  });

  it.each`
    index | expectedPosition | status
    ${0}  | ${30}            | ${"in_transit"}
    ${1}  | ${55}            | ${"incoming"}
    ${2}  | ${70}            | ${"stopped"}
  `(
    "positions vehicles according to status $status",
    ({ index, expectedPosition }) => {
      const node = wrapper.find(".m-schedule-diagram__vehicle").at(index);
      const { top } = node.get(0).props.style; // e.g. "30px"
      const top_number = parseInt(top.substring(0, 2)); // e.g. 30
      expect(top_number).toEqual(expectedPosition);
    }
  );

  it("displays the vehicle's tooltip_text", () => {
    for (let index of [0, 1, 2]) {
      const node = wrapper.find(".m-schedule-diagram__vehicle").at(index);
      expect(tooltipText(node)).toContain(vehicles[index].tooltip_text);
    }
  });
});

it("VehicleIcons does not render if store has no data for the relevant stop id", () => {
  const wrapperWithoutStopData = mount(
    <redux.Provider store={store}>
      <VehicleIcons stop_id={"test-stop-other"} vehicles={other_vehicles} />
    </redux.Provider>
  );
  expect(wrapperWithoutStopData.find(VehicleIcons).html()).toBeNull();
});

describe("LiveVehicles", () => {
  it("does not render if the channel returns no vehicles", () => {
    // mock useVehicleMarkersChannel to return no vehicles
    jest.spyOn(UseVehicleChannel, "default").mockImplementation(() => []);
    const wrapper = mount(
      <redux.Provider store={store}>
        <LiveVehicles channel="vehicles:MyRoute:0" />
      </redux.Provider>
    );
    expect(wrapper.find(LiveVehicles).html()).toBeNull();
  });

  it("does not render if the channel returns vehicles with no stop ids", () => {
    const vehiclesWithoutStops = cloneDeep(vehicles).map(v => {
      delete v.stop_id;
      return v;
    });
    // mock useVehicleMarkersChannel to return vehicles without stop_id
    jest
      .spyOn(UseVehicleChannel, "default")
      .mockImplementation(() => vehiclesWithoutStops);
    const wrapper = mount(
      <redux.Provider store={store}>
        <LiveVehicles channel="vehicles:MyRoute:0" />
      </redux.Provider>
    );
    expect(wrapper.find(LiveVehicles).html()).toBeNull();
  });

  it("renders vehicles if the channel returns vehicles", () => {
    // mock useVehicleMarkersChannel to return vehicles
    jest.spyOn(UseVehicleChannel, "default").mockImplementation(() => vehicles);
    const wrapper = mount(
      <redux.Provider store={store}>
        <LiveVehicles channel="vehicles:MyRoute:0" />
      </redux.Provider>
    );
    expect(wrapper.find(LiveVehicles).html()).not.toBeNull();
    expect(wrapper.find(VehicleIcons).exists()).toBeTruthy();
    expect(wrapper.find(VehicleIcons).prop("stop_id")).toEqual(stop.id);
    expect(wrapper.find(LiveVehicles).debug()).toMatchSnapshot();
    const vehicleNodes = wrapper
      .find(VehicleIcons)
      .find(".m-schedule-diagram__vehicle");
    expect(vehicleNodes).toHaveLength(vehicles.length);
  });
});
