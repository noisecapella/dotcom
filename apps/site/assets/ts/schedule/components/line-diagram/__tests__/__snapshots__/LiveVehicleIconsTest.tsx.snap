// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveVehicles renders vehicles if the channel returns vehicles 1`] = `
"<LiveVehicles channel=\\"vehicles:MyRoute:0\\">
  <VehicleIcons stop_id=\\"test-stop\\" vehicles={{...}}>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
  </VehicleIcons>
</LiveVehicles>"
`;

exports[`VehicleIcons renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <VehicleIcons stop_id=\\"test-stop\\" vehicles={{...}}>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 1</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 2</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\"> tooltip text for vehicle 3</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
  </VehicleIcons>
</Provider>"
`;
